src/audio/types.h

```diff
@@ -0,0 +1,188 @@
+    static constexpr SINT kValueMax = 255; // upper bound (inclusive, 8-bit unsigned integer)
+
+    static constexpr ChannelCount min() {
+        return ChannelCount(kValueMin);
+    }
+    static constexpr ChannelCount max() {
+        return ChannelCount(kValueMax);
+    }
+
+    static ChannelCount fromLayout(ChannelLayout layout) {
+        switch (layout) {
+        case ChannelLayout::Mono:
+            return ChannelCount(1);
+        case ChannelLayout::DualMono:
+            return ChannelCount(1);
+        case ChannelLayout::Stereo:
+            return ChannelCount(2);
+        }
+        DEBUG_ASSERT(!"unreachable code");
+    }
+
+    explicit constexpr ChannelCount(SINT value = kValueDefault)
+            : m_value(value) {
+    }
+    explicit ChannelCount(ChannelLayout layout)
+            : m_value(fromLayout(layout).m_value) {
+    }
+
+    constexpr bool isValid() const {
+        return (kValueMin <= m_value) &&
+                (m_value <= kValueMax);
+    }
+
+    /*implicit*/ constexpr operator SINT() const {
+        return m_value;
+    }
+
+  private:
+    SINT m_value;
+};
+
+// Defines the ordering of how samples from multiple channels are
+// stored in contiguous buffers:
+//    - Planar: Channel by channel
+//    - Interleaved: Frame by frame
+// The samples from all channels that are coincident in time are
+// called a "frame" (or more specific "sample frame").
+//
+// Example: 10 stereo samples from left (L) and right (R) channel
+// Planar layout:      LLLLLLLLLLRRRRRRRRRR
+// Interleaved layout: LRLRLRLRLRLRLRLRLRLR
+enum class SampleLayout {
+    Planar,
+    Interleaved
+};
+
+typedef std::optional<SampleLayout> OptionalSampleLayout;
+
+QDebug operator<<(QDebug dbg, SampleLayout arg);
+
+class SampleRate {
+  private:
+    static constexpr SINT kValueDefault = 0;
+
+  public:
+    static constexpr SINT kValueMin = 8000;   // lower bound (inclusive, = minimum MP3 sample rate)
+    static constexpr SINT kValueMax = 192000; // upper bound (inclusive)
+
+    static constexpr SampleRate min() {
+        return SampleRate(kValueMin);
+    }
+    static constexpr SampleRate max() {
+        return SampleRate(kValueMax);
+    }
+
+    static constexpr const char* unit() {
+        return "Hz";
+    }
+
+    explicit constexpr SampleRate(SINT value = kValueDefault)
+            : m_value(value) {
+    }
+
+    constexpr bool isValid() const {
+        return (kValueMin <= m_value) &&
+                (m_value <= kValueMax);
+    }
+
+    /*implicit*/ constexpr operator SINT() const {
+        return m_value;
+    }
+
+  private:
+    SINT m_value;
+};
+
+QDebug operator<<(QDebug dbg, SampleRate arg);
+
+// The bitrate is measured in kbit/s (kbps) and provides information
+// about the level of compression for lossily encoded audio streams.
+// It depends on the metadata and decoder if a value for the bitrate
+// is available, i.e. it might be invalid if it cannot be determined.
+class Bitrate {
```

Comment:
Should this class have a way to express a variable bitrate? Invalid is not the same as variable.

Reply:
This is always interpreted as _average_ bitrate. We could rename the type or the member in `StreamInfo` The bitrate is only informational and its interpretation closely depends on the actual format.

Reply:
Since we use the term _bitrate_ in the meaning of _average bitrate_ all over I would prefer to leave everything as is.

Reply:
How about making a note of this in the comment here?

Reply:
Agreed and done.

Reply:
I have just documented the status quo.

The (invalid) default values are required for the Qt metatype system. This doesn't need to be mentioned separately. The compiler will stop you from removing default constructors of non-optional values.

Reply:
okay